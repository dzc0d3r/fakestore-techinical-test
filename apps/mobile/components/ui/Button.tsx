import { AntDesign } from "@expo/vector-icons";
import React from "react";
import {
  ActivityIndicator,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
} from "react-native";

type ButtonProps = {
  label: string;
  onPress: () => void;
  variant?: "primary" | "secondary" | "danger" | "outline";
  icon?: React.ReactNode;
  loading?: boolean;
  style?: ViewStyle;
};

export const Button: React.FC<ButtonProps> = ({
  label,
  onPress,
  variant = "primary",
  icon,
  loading = false,
  style,
}) => {
  const getVariantStyle = (): ViewStyle => {
    switch (variant) {
      case "secondary":
        return { backgroundColor: "#4CAF50", borderColor: "#45a049" };
      case "danger":
        return { backgroundColor: "#dc3545", borderColor: "#c82333" };
      case "outline":
        return {
          backgroundColor: "transparent",
          borderWidth: 1,
          borderColor: "#007bff",
        };
      default: // primary
        return { backgroundColor: "#2196F3", borderColor: "#1976D2" };
    }
  };

  const getTextColor = (): TextStyle => {
    return variant === "outline" ? { color: "#007bff" } : { color: "white" };
  };

  return (
    <TouchableOpacity
      style={[styles.button, getVariantStyle(), style]}
      onPress={onPress}
      disabled={loading}
    >
      {loading ? (
        <ActivityIndicator color={getTextColor().color} />
      ) : (
        <React.Fragment>
          {icon && <>{icon}</>}
          {icon && <Text style={{ width: 8 }} />}
          <Text style={[styles.text, getTextColor()]}>{label}</Text>
        </React.Fragment>
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    paddingVertical: 12,
    paddingHorizontal: 16,
    borderRadius: 8,
    borderWidth: 1,
  },
  text: {
    fontSize: 16,
    fontWeight: "500",
  },
});
