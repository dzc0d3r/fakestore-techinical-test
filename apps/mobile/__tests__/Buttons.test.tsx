import { AntDesign } from "@expo/vector-icons";
import React from "react";
import renderer from "react-test-renderer";
import { Button } from "@/components/ui/Button";

describe("Button Component", () => {
  const mockPress = jest.fn();

  it("renders primary button correctly", () => {
    const tree = renderer
      .create(<Button label="Primary" onPress={mockPress} variant="primary" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders secondary button correctly", () => {
    const tree = renderer
      .create(
        <Button label="Secondary" onPress={mockPress} variant="secondary" />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders danger button correctly", () => {
    const tree = renderer
      .create(<Button label="Danger" onPress={mockPress} variant="danger" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders outline button correctly", () => {
    const tree = renderer
      .create(<Button label="Outline" onPress={mockPress} variant="outline" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });


  it("renders loading state correctly", () => {
    const tree = renderer
      .create(<Button label="Loading" onPress={mockPress} loading={true} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with custom style correctly", () => {
    const tree = renderer
      .create(
        <Button
          label="Custom Style"
          onPress={mockPress}
          style={{ marginTop: 10, width: 200 }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders primary button with icon and loading state correctly", () => {
    const tree = renderer
      .create(
        <Button
          label="Complete"
          onPress={mockPress}
          variant="primary"
          icon={<AntDesign name="check" size={16} color="white" />}
          loading={true}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
