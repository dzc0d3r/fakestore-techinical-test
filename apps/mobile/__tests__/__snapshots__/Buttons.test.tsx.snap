// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders danger button correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#dc3545",
      "borderColor": "#c82333",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "500",
        },
        {
          "color": "white",
        },
      ]
    }
  >
    Danger
  </Text>
</View>
`;

exports[`Button Component renders loading state correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2196F3",
      "borderColor": "#1976D2",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <ActivityIndicator
    color="white"
  />
</View>
`;

exports[`Button Component renders outline button correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#007bff",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "500",
        },
        {
          "color": "#007bff",
        },
      ]
    }
  >
    Outline
  </Text>
</View>
`;

exports[`Button Component renders primary button correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2196F3",
      "borderColor": "#1976D2",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "500",
        },
        {
          "color": "white",
        },
      ]
    }
  >
    Primary
  </Text>
</View>
`;

exports[`Button Component renders primary button with icon and loading state correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2196F3",
      "borderColor": "#1976D2",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <ActivityIndicator
    color="white"
  />
</View>
`;

exports[`Button Component renders secondary button correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4CAF50",
      "borderColor": "#45a049",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "500",
        },
        {
          "color": "white",
        },
      ]
    }
  >
    Secondary
  </Text>
</View>
`;

exports[`Button Component renders with custom style correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2196F3",
      "borderColor": "#1976D2",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginTop": 10,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "width": 200,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "500",
        },
        {
          "color": "white",
        },
      ]
    }
  >
    Custom Style
  </Text>
</View>
`;
